### ISSUE TITLE
FR: Eager evaluation of macros in macro definitions

### ISSUE THREAD

**Issue by Slayer95** [wesnoth/wesnoth#7522]
### Describe the desired feature

Let ``src.cfg`` be

```
#define DMSG TXT
{DEBUG_MSG {TXT}}
#enddef

#define HELLO_WORLD
{DMSG "Hello"}
{DMSG "World"}
#enddef

#undef DMSG

{HELLO_WORLD}
```

As of 1.16.8, parse fails with `Macro/file 'DMSG' is missing`. I figure this happens
because HELLO_WORLD is expanded only after being called. However, it could be evaluated
as soon as it is defined, so that ``src.cfg`` becomes equivalent to.

```
#define DMSG TXT
{DEBUG_MSG {TXT}}
#enddef

#define HELLO_WORLD
{DEBUG_MSG "Hello"}
{DEBUG_MSG "World"}
#enddef

#undef DMSG

{HELLO_WORLD}
```

**Comment by Pentarctagon** [comment]
Why is the undef needed?

**Comment by Slayer95** [comment]
> Why is the undef needed?

1. It avoids pollution of global namespace.
2. It allows static analysis tools to quickly conclude that ``DMSG`` is a local macro called exactly twice with certain parameters.
3. For a practical use case, if ``DMSG`` contains a translatable ``_ "String with {MACRO}"``, ``wmlxgettext`` could be made aware of all its variants. Without ``#undef``, solving the problem of translatable strings with macros would have to fall back to a slow path.

**Comment by CelticMinstrel** [comment]
So related to #1135?

**Comment by Slayer95** [comment]
> So related to https://github.com/wesnoth/wesnoth/issues/1135?

#1135 seems to be related to variable substitution at the ``gettext`` level, while this issue is about macro substitution before ``gettext`` sees any input. While macros could be also used to solve the plurals issue for the simplest cases and languages, such an usage would not be recommended.

On the other hand, their main application in i18n would lie in providing ideal solutions/alternatives for https://github.com/wesnoth/wesnoth/pull/6165 and similar scenarios with text variants orthogonal to grammar.

**Comment by Pentarctagon** [comment]
This sounds like it'd also be a compatibility breaking change, since now the order of macro definition would be dependent on more than where the macro is used.

**Comment by Slayer95** [comment]
> This sounds like it'd also be a compatibility breaking change, since now the order of macro definition would be dependent on more than just where the macro is used.

If ``HELLO_WORLD`` is loaded before ``DMSG``, it could then still be lazily evaluated with no compat breakage. In that case, ``#undef`` would not be usable, just like today.

Oddly, enough, the following code already does *not* work (``Macro/file 'TXT' is missing``), so there wouldn't be compat breakage here either.
```c++
#define DMSG TXT
{DEBUG_MSG {TXT}}
#enddef

#define HELLO_WORLD
{DMSG "Hello"}
{DMSG "World"}
#enddef

#undef DMSG
#def DMSG TXT
{DEBUG_MSG "Monkey"}
{DEBUG_MSG {TXT}}
{DEBUG_MSG "Patch"}
#enddef

{HELLO_WORLD}
```

**Comment by CelticMinstrel** [comment]
Is that the exact code you used? I think if you fix your syntax error (`#def DMSG TXT`) it would actually work.

**Comment by Slayer95** [comment]
> Is that the exact code you used? I think if you fix your syntax error (`#def DMSG TXT`) it would actually work.

Oh! Good catch. Yeah, so, that means there <i>is</i> breakage there after all. No idea how widespread that pattern is.

**Comment by ProditorMagnus** [comment]
Purpose of macro is to disguise long text from user. When you want to do something more advanced, consider Lua.

> It allows static analysis tools to quickly conclude that DMSG is a local macro called exactly twice with certain parameters.
I consider macros meaningless for analyzing, instead running preprocessor to get rid of them and analyze resulting WML.

**Comment by ProditorMagnus** [comment]
Macro is irrelevant for tools, provided preprocessor bugs are fixed.

**Comment by Slayer95** [comment]
> Macro is irrelevant for tools, provided preprocessor bugs are fixed. I mean those like https://github.com/wesnoth/wesnoth/issues/7471 https://github.com/wesnoth/wesnoth/issues/4698 https://github.com/wesnoth/wesnoth/issues/4697

Setting aside the point that requiring all static analysis to use the builtin compiler (and to have BfW preinstalled) is not reasonable, the reason why local macros are useful is that they make analysis of _a single file_ more fruitful, be it through _preprocessor_ or alternate means. On the other hand, if all macros are global, it will often be needed to analyze the entire addon before reporting on any individual event/macro/file whatever.

**Comment by ProditorMagnus** [comment]
Would moving all #undef to end of file solve your use case?

**Comment by Pentarctagon** [comment]
What static analysis are you referring to?

**Comment by Slayer95** [comment]
> What static analysis are you referring to?

2. It allows static analysis tools to quickly conclude that DMSG is a local macro called exactly twice with certain parameters.
3. For a practical use case, if DMSG contains a translatable _ "String with {MACRO}", wmlxgettext could be made aware of all its variants. Without #undef, solving the problem of translatable strings with macros would have to fall back to a slow path.

> any references within it to a known macro is substituted (recursively), while references to unknown macros are left in place

Beautifully put. However, C doesn't have recursive expand-on-define behavior. That's a feature of [hygienic macros](https://en.wikipedia.org/wiki/Hygienic_macro), seen in languages such as Scheme, Clojure, and Racket.

In particular, [Racket supports both styles](https://beautifulracket.com/explainer/hygiene.html) through different syntaxes. Racket assumes hygiene by default, yet allows defining which identifiers get evaluated lazily. In contrast, current Wesnoth behavior is unhygienic. Either the preprocessor can be changed to optimistically do recursive expansion and fall back to leaving undefined references in place, or the gap could be bridged by using some ``#capture IDENTIFIER`` directive inside the macro definition.

**Comment by Pentarctagon** [comment]
@Slayer95 yes, but what static analysis tools are you talking about? As far as I know, no such tools exist.

**Comment by Slayer95** [comment]
> It allows static analysis tools to quickly conclude that DMSG is a local macro called exactly twice with certain parameters.

``wmltools3`` does that. But its results should be non-conclusive for non-hygienic macros.

> For a practical use case, if DMSG contains a translatable _ "String with {MACRO}", wmlxgettext could be made aware of all its variants.

WIP

**Comment by Pentarctagon** [comment]
Also, for compatibility breakage, I mean something like this which works today:
```
#define AB
{A}{B}#enddef

#define A
a#enddef

#define B
b#enddef

{AB}
```
would no longer work. I could see that being something implicitly relied on in current code.

**Comment by Slayer95** [comment]
```
#define AB
{A}{B}#enddef

#define A
a#enddef

#define B
b#enddef

{AB}
```

This is where the double-pass is relevant. After ``#define AB`` fails to find ``A`` nor ``B``, the macro is marked so that ``{AB}`` evaluates them on runtime.

**Comment by Pentarctagon** [comment]
What do you mean by runtime?

**Comment by Slayer95** [comment]
> What do you mean by runtime?

... Whenever they are evaluated today.

**Comment by CelticMinstrel** [comment]
At macro substitution time.

I'm not 100% but I think "recursive substitution" is only achieved by dumping the macro's definition into the stream and then parsing it like normal.
